implementations/rust/performance-guideline.md

```diff
@@ -0,0 +1,344 @@
+# Rust Performance Guidelines
+
+## Intro
+
+This is a quick guideline to avoid the most common performance pitfalls in rust.
+Although performance is important, it's not the only thing to consider when writing code. This guideline should be closely applied only in hot-paths, where performance is crucial, and more loosely applied in cold-paths, where readability and maintainability are more important.
+
+## Memory Allocations
+Frequent memory allocations can degrade performance either for the sheer number of allocations or for the consequent memory fragmentation. Most of the guidelines here are to avoid unnecessary allocations and copies. For Ockam Command, the typical memory fragmentation symptoms are unstable or degraded performance.
+
+## Tools
+There are some tools that can help you to profile Ockam Command portals. See [here](tools/profile/README.md) for more information.
+
+## Guidelines
+
+### Owned Parameters
+If the function needs an owned copy of the element, ask an owned instance directly in the parameter. This will allow the caller to decide if it wants to clone the element or give its own.
+
+```rust
+// Bad
+fn example(item: &Type) {
+  self.item = item.clone();
+}
+
+// Good
+fn example(item: Type) {
+  self.item = item;
+}
+```
+
+The same can be applied to constructors, if the instance needs to own the element, ask for an owned element in the constructor.
+
+```rust
+// Bad
+impl Type {
+  fn new(item: &Type) -> Self {
+    Self {
+      item: item.clone(),
+    }
+  }
+}
+
+// Good
+impl Type {
+  fn new(item: Type) -> Self {
+    Self {
+      item,
+    }
+  }
+}
+```
+
+### Setters returns references
```

Comment:
I haven't noticed that pattern anywhere before. I used to either return nothing, or a reference to self so that different setter calls could be chained. What's the idea behind returning the reference to the value being set?

Reply:
Most of the time you only need a reference to an object, but if you use it in a setter you "lose it".
Now you are faced with two choices, either you add a getter or clone the object before passing it to the setter.

Reply:
Do we have an example where you need a reference to an object? If it's not common, I would say we can just use a getter to get the reference